### PR TITLE
[AMD] Enable share expert fusion with router experts for Qwen3.5 BF16 & FP8

### PR DESCRIPTION
**Pull request by zhentaocc**
## Motivation

Qwen2 MoE and Qwen3.5 MoE models use a **shared expert** in addition to routed experts. When `shared_expert_intermediate_size == moe_intermediate_size`, the shared expert can be fused with routed experts so that each token attends to top-k routed experts plus one shared expert (topk+1) in a single MoE dispatch, reducing kernel launches and improving inference efficiency. This PR adds shared expert fusion support for Qwen2 MoE (when using Aiter on ROCm/HIP) and improves Qwen3.5 MoE weight loading to correctly handle the fused shared expert layout.

### `python/sglang/srt/models/qwen2_moe.py`

- **`_determine_num_fused_shared_experts()`**: New helper that returns 1 when shared expert fusion is enabled (requires `shared_expert_intermediate_size == moe_intermediate_size`, not disabled via `--disable-shared-experts-fusion`, and `SGLANG_USE_AITER=1` on HIP).
- **`_get_shared_expert_weights()`**: Returns `sigmoid(shared_expert_gate(hidden_states))` for the fused shared expert weights.
- **`_append_shared_to_topk_output()`**: Appends shared expert IDs and weights to the top-k output before the fused MoE forward.
- **`_forward_router_experts()`**: After top-k selection on gate logits, appends shared expert via `_append_shared_to_topk_output()` when fusion is enabled.
- **Experts and TopK**: `top_k` and `num_experts` now include `num_fused_shared_experts` when fusion is active.

### `python/sglang/srt/models/qwen3_5.py`

- **`_get_num_fused_shared_experts()`**: New helper used by `Qwen3_5MoeForConditionalGeneration` to obtain `num_fused_shared_experts` from the first layer’s MLP.
- **load_weights (Qwen3_5MoeForConditionalGeneration)**: 
  - Same `num_experts` adjustment for expert params mapping.
  - Remaps `mlp.shared_expert.*` to `mlp.experts.{num_experts_base}.*` when fusion is enabled.
  - Adds `fused_expert_params_mapping` entries for shared expert (`gate_proj`, `up_proj`, `down_proj`, and combined `gate_up_proj`).
  - Handles both separate (`gate_proj`/`up_proj`) and combined (`gate_up_proj`) checkpoint layouts for the shared expert.

## Accuracy Tests
Model:  Qwen/Qwen3.5-397B-A17B
enable fusion:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9682|±  |0.0048|
|     |       |strict-match    |     5|exact_match|↑  |0.9697|±  |0.0047|

disable fusion:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9697|±  |0.0047|
|     |       |strict-match    |     5|exact_match|↑  |0.9712|±  |0.0046|

## Benchmarking 
Before
```bash
 sglang serve     --attention-backend triton     --model-path $MODEL     --host=0.0.0.0     --port $PORT     --tensor-parallel-size $TP     --trust-remote-code     --mem-fraction-static 0.8 
```
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     160       
Benchmark duration (s):                  192.69    
Total input tokens:                      147038    
Total input text tokens:                 147038    
Total generated tokens:                  147952    
Total generated tokens (retokenized):    147623    
Request throughput (req/s):              0.83      
Input token throughput (tok/s):          763.08    
Output token throughput (tok/s):         767.82    
Peak output token throughput (tok/s):    864.00    
Peak concurrent requests:                21        
Total token throughput (tok/s):          1530.90   
Concurrency:                             15.39     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18535.68  
Median E2E Latency (ms):                 18662.69  
P90 E2E Latency (ms):                    20331.16  
P99 E2E Latency (ms):                    20857.56  
---------------Time to First Token----------------
Mean TTFT (ms):                          128.35    
Median TTFT (ms):                        97.52     
P99 TTFT (ms):                           409.59    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          19.92     
Median TPOT (ms):                        20.06     
P99 TPOT (ms):                           20.38     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           19.93     
Median ITL (ms):                         18.77     
P95 ITL (ms):                            19.03     
P99 ITL (ms):                            99.89     
Max ITL (ms):                            405.48    
==================================================
```
After
```

============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     160       
Benchmark duration (s):                  184.30    
Total input tokens:                      147038    
Total input text tokens:                 147038    
Total generated tokens:                  147952    
Total generated tokens (retokenized):    147620    
Request throughput (req/s):              0.87      
Input token throughput (tok/s):          797.83    
Output token throughput (tok/s):         802.79    
Peak output token throughput (tok/s):    896.00    
Peak concurrent requests:                21        
Total token throughput (tok/s):          1600.62   
Concurrency:                             15.39     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   17732.18  
Median E2E Latency (ms):                 17863.97  
P90 E2E Latency (ms):                    19450.08  
P99 E2E Latency (ms):                    19969.85  
---------------Time to First Token----------------
Mean TTFT (ms):                          124.12    
Median TTFT (ms):                        93.28     
P99 TTFT (ms):                           406.08    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          19.06     
Median TPOT (ms):                        19.18     
P99 TPOT (ms):                           19.49     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           19.06     
Median ITL (ms):                         17.94     
P95 ITL (ms):                            18.25     
P99 ITL (ms):                            95.79     
Max ITL (ms):                            414.58    
==================================================

```
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/qwen2_moe.py [hunk 1]
```diff
@@ -162,17 +167,30 @@ def __init__(
...
                 f"Tensor parallel size {self.tp_size} is greater than "
                 f"the number of experts {config.num_experts}."
             )
 
+        self.num_fused_shared_experts = self._determine_num_fused_shared_experts(
+            config,
+            quant_config,
+        )
+        if self.num_fused_shared_experts > 0:
+            rank0_log(
```
Comment (hubertlu-tw) [inline python/sglang/srt/models/qwen2_moe.py]: Do we still need this? I assume this is for debugging purposes. Same for other places.

File: python/sglang/srt/models/qwen2_moe.py [hunk 2]
```diff
@@ -296,9 +371,14 @@ def _forward_deepep(self, hidden_states: torch.Tensor, forward_batch: ForwardBat
         return final_hidden_states
 
     def _forward_router_experts(self, hidden_states: torch.Tensor):
-        # router_logits: (num_tokens, n_experts)
-        router_logits, _ = self.gate(hidden_states)
-        topk_output = self.topk(hidden_states, router_logits)
+        gate_logits, _ = self.gate(hidden_states)
```
Comment (yichiche) [inline python/sglang/srt/models/qwen2_moe.py]: @zhentaocc Is there a specific reason for changing the name from router_logits to gate_logits? If not, we should keep it as is to avoid unnecessary changes.

File: python/sglang/srt/models/qwen2_moe.py [hunk 3]
```diff
@@ -891,6 +973,7 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                     if weight_name not in name:
                         continue
                     name = name.replace(weight_name, param_name)
+                    rank0_log(f"load_weights: {raw_weight_name} -> {name}")
```
Comment (yichiche) [inline python/sglang/srt/models/qwen2_moe.py]: We should remove this logging code from the production environment.

File: python/sglang/srt/models/qwen2_moe.py [hunk 4]
```diff
@@ -234,6 +257,20 @@ def __init__(
...
+    def _determine_num_fused_shared_experts(self, config: PretrainedConfig, quant_config: Optional[QuantizationConfig]) -> int:
+        """Determine if shared expert can be fused with router experts (topk+1).
+        Fusion requires shared_expert_intermediate_size == moe_intermediate_size,
+        support_shared_expert_fusion=True (e.g. Qwen3.5 MoE), and Aiter on HIP.
+        """
+        if (not self.support_shared_expert_fusion
+            or get_global_server_args().disable_shared_experts_fusion is True
+            or getattr(config, "shared_expert_intermediate_size", 0) <= 0
+            or config.shared_expert_intermediate_size != config.moe_intermediate_size
+            or not _use_aiter
```
Comment (yichiche) [inline python/sglang/srt/models/qwen2_moe.py]: Add `or get_moe_a2a_backend().is_deepep()`  to check if deepep is enable.
Otherwise, self._forward_shared_experts in the _forward_deepep function will be overridden because self.shared_expert is set to None.

File: python/sglang/srt/models/qwen3_5.py [hunk 5]
```diff
@@ -1106,13 +1117,17 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
...
+
         # Params for weights, fp8 weight scales, fp8 activation scales
         # (param_name, weight_name, expert_id, shard_id)
         expert_params_mapping = FusedMoE.make_expert_params_mapping(
             ckpt_gate_proj_name="gate_proj",
             ckpt_down_proj_name="down_proj",
             ckpt_up_proj_name="up_proj",
-            num_experts=self.config.num_experts,
+            num_experts=num_experts,
         )
```
Comment (yichiche) [inline python/sglang/srt/models/qwen3_5.py:1168]: You may also need fused_expert_params_mapping here? Will add num_experts be sufficient?

File: python/sglang/srt/models/qwen3_5.py [hunk 6]
```diff
@@ -1164,6 +1177,7 @@ def load_fused_expert_weights(
         params_dict = dict(self.named_parameters(remove_duplicate=False))
 
         for name, loaded_weight in weights:
+            raw_weight_name = name
```
Comment (yichiche) [inline python/sglang/srt/models/qwen3_5.py]: We should remove this logging code from the production environment.

File: python/sglang/srt/models/qwen3_5.py (hunk 5 above)
Comment (zhentaocc) [reply python/sglang/srt/models/qwen3_5.py:1168]: The existing `fused_expert_params_mapping ` is for gate_up_proj fusion,  not for share expert fusion.

For shared expert, it will be mapped as `experts.512.gate_up_proj` or `experts.512.down_proj`. It's handled by
```python
                            if f"{num_experts_base}.gate_up_proj" in name:
                                loaded_weight = loaded_weight.chunk(2, dim=-2)
                                weight_loader(
                                    param,
                                    loaded_weight[0],
                                    name_mapped,
                                    "w1",
                                    expert_id,
                                )
                                weight_loader(
                                    param,
                                    loaded_weight[1],
                                    name_mapped,
                                    "w3",
                                    expert_id,
                                )
                            else:
                                weight_loader(
                                    param,
                                    loaded_weight,
                                    name_mapped,
                                    shard_id,
                                    expert_id,
                                )
```
split out gate and up, and load into w1 and w3.

If keys in ckpt in format of `gate, up, down`, it would be sufficient adding num_experts to 512.

Comment (zhentaocc) [reply python/sglang/srt/models/qwen3_5.py:1168]: Oh, I just realize no need to modify `load_weights` for this class. I am restricting the changes to influence `Qwen3_5ForConditionalGeneration` only. So removed these changes.

**Comment by yichiche** [comment]
@zhentaocc Please fix lint issue, I will kick off CI again once it's done

File: python/sglang/srt/models/qwen2_moe.py [hunk 7]
```diff
@@ -234,6 +252,25 @@ def __init__(
...
+        Fusion requires shared_expert_intermediate_size == moe_intermediate_size,
+        support_shared_expert_fusion=True (e.g. Qwen3.5 MoE), and Aiter on HIP.
+        """
+        if (
+            not self.support_shared_expert_fusion
+            or get_global_server_args().disable_shared_experts_fusion is True
+            or getattr(config, "shared_expert_intermediate_size", 0) <= 0
+            or config.shared_expert_intermediate_size != config.moe_intermediate_size
+            or not _use_aiter
+            or quant_config is not None
```
Comment (zhentaocc) [inline python/sglang/srt/models/qwen2_moe.py]: This PR only works on BF16 for now for easier review. And I also need some time to resolve accuracy issue on FP8 weight scale loading. I can raise new PR to address this for FP8 and remove line 268.  @HaiShaw @yichiche

Comment (zhentaocc) [reply python/sglang/srt/models/qwen2_moe.py]: additional works we might collaborate on:

- [x] FP8/MXFP4 support
- [ ] Qwen35 has separate shared_gate, which I also try to fuse it with gate_proj.

Comment (yichiche) [reply python/sglang/srt/models/qwen2_moe.py]: FP8 accuracy issue identified, will need aiter upgrade to fix split_k issue.

**Review by HaiShaw — CHANGES_REQUESTED** [review]
Refactoring is needed

**Comment by HaiShaw** [comment]
@amd-bot ci-status

**Review by HaiShaw — APPROVED** [review]

**Review by HaiShaw — APPROVED** [review]
